cpp/src/io/csv/csv_writer.cu

```diff
@@ -0,0 +1,284 @@
+ */
+
+/**
+ * @file csv_writer.cu  code to create csv file
+ *
+ * CSV Writer
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <cuda_runtime.h>
+#include <nvstrings/NVStrings.h>
+#include <rmm/rmm.h>
+#include <rmm/thrust_rmm_allocator.h>
+#include "cudf.h"
+#include "utilities/error_utils.hpp"
+
+
+// called by write_csv method
+NVStrings* column_to_strings_csv( gdf_column* col, const char* delimiter, const char* strue, const char* sfalse )
+{
+    NVStrings* rtn = nullptr;
+    unsigned int rows = (unsigned int)col->size;
+    unsigned char* valid = (unsigned char*)col->valid;
+    switch( col->dtype )
+    {
+        case GDF_STRING:
+            rtn = ((NVStrings*)col->data)->copy();
+            break;
+        case GDF_INT8:
+            rtn = NVStrings::create_from_bools((const bool*)col->data,rows,strue,sfalse,valid);
+            break;
+        case GDF_INT32:
+            rtn = NVStrings::itos((const int*)col->data,rows,valid);
+            break;
+        case GDF_INT64:
+            rtn = NVStrings::ltos((const long*)col->data,rows,valid);
+            break;
+        case GDF_FLOAT32:
+            rtn = NVStrings::ftos((const float*)col->data,rows,valid);
+            break;
+        case GDF_FLOAT64:
+            rtn = NVStrings::dtos((const double*)col->data,rows,valid);
+            break;
+        case GDF_DATE64:
+            rtn = NVStrings::long2timestamp((const unsigned long*)col->data,rows,NVStrings::seconds,valid);
+            break;
+        default:
+            break; // should not happen
+    }
+    if( !rtn )
+        return nullptr;
+
+    // probably could collapse this more
+    bool bquoted = (col->dtype==GDF_STRING);
+    // check for delimeters and quotes
+    bool* bmatches = nullptr;
+    RMM_TRY( RMM_ALLOC(&bmatches,rows*sizeof(bool),0) );
+    if( rtn->contains("\"",bmatches) > 0 )
+    {
+        NVStrings* esc = rtn->replace("\"","\"\"");
+        NVStrings::destroy(rtn);
+        rtn = esc;
+    }
+    else if( rtn->contains(",",bmatches) > 0 )
+        bquoted = true;
+    RMM_TRY( RMM_FREE( bmatches, 0 ) );
+    if( bquoted )
+    {
+        // prepend and append quotes
+        NVStrings* pre = rtn->slice_replace("\"",0,0);
+        NVStrings::destroy(rtn);
+        rtn = pre->slice_replace("\"",-1,-1);
+        NVStrings::destroy(pre);
+    }
+    if( delimiter && *delimiter )
+    {
+        NVStrings* dstr = rtn->slice_replace(delimiter,-1,-1);
+        NVStrings::destroy(rtn);
+        rtn = dstr;
+    }
+    return rtn;
+}
+
+//
+// The args structure is interpretted as documented in io_types.h
+// This will create a CSV format by allocating host memory for the
+// entire output and determine pointers for each row/column entry.
+// Each column is converted to an NVStrings instance and then
+// copied into their position in the output memory. This way,
+// one column is processed at a time minimizing device memory usage.
+//
+gdf_error write_csv(csv_write_arg* args)
```

Comment:
Sorry, I don't understand this comment. I was trying to match the style of read_csv. I guess I need specific help on what needs to be changed here.

Reply:
Discussed with Jake. To keep the interface between read_csv and write_csv consistent, it's OK for this PR to not take on the C++ API as described in the gist. But we will be converting to C++ APIs so it will need to happen in a later PR (probably together with read_csv).